### PR TITLE
gh-99761: add invalid_index macro

### PR DESCRIPTION
**Pull request by eendebakpt**
In `listobject.c` there is an [optimization](https://github.com/python/cpython/blob/main/Objects/listobject.c#L230) to check whether an index is valid (e.g. `0 <= index < N`) using a single comparison. The cast-to-unsigned optimization is used in current main in 3 locations: `tupleobject.c`, `_collectionsmodule.c` and `bytesobject.c`.

It is a micro optimization, but the code generated is different than the plain ` i < 0 || i >= Py_SIZE(a)`. This PR tries to:

i) apply the optimization to more locations
ii) make the code more consistent

By replacing index checks with a single method `_Py_is_valid_index` that includes the optimization we have consistency in the code and have the optimized check for all index checks.

* Issue: gh-99761

**Notes:**

* The new method is private and is static inline (https://peps.python.org/pep-0670/)
* In `bytecodes.c`  the optimized expression is used in three places. E.g.
```
DEOPT_IF(((size_t)signed_magnitude) > 1, BINARY_SUBSCR);
```
With the `_Py_is_valid_index`  method it looks like
```
DEOPT_IF(!_Py_is_valid_index(signed_magnitude, 2), BINARY_SUBSCR);
```
which is in my opinion not very readable. For these cases a separate PR was created: #100064

### CODE REVIEW

**Comment by sweeneyde** [comment]
I think this is generally a good idea, but I'm not sure this should go in the public API. If it does, its name should probably begin with 'Py...'. See https://devguide.python.org/developer-workflow/c-api/ and  https://peps.python.org/pep-0007/#naming-conventions .

cc @vstinner

**Comment by eendebakpt** [comment]
> I think this is generally a good idea, but I'm not sure this should go in the public API. If it does, its name should probably begin with 'Py...'. See https://devguide.python.org/developer-workflow/c-api/ and https://peps.python.org/pep-0007/#naming-conventions .
> 
> cc @vstinner

Agreed. What would be a good location for the private version? One of the `Include/internal/python_xxxx` files?

**Comment by rhettinger** [comment]
* Keep the code as `valid_index` rather than `invalid_index` so that we don't even up with double negatives.

* Please leave the collections.deque code alone.  The author (me) aspires to keep this code not tightly coupled to the rest of the C API. The code is currently clean an doesn't need to be changed.

* Are you *certain* that inline functions always get inlined?  We know that macros do *always* get inlined.  The `valid_index` macro currently gets used in performance critical code.

* As Dennis says, this should not be part of the public API.  It is a performance trick that happens to currently be useful. As compilers get better, this would be done automatically and we would prefer the simple inline code rather than the extra layer of abstraction.

**Comment by eendebakpt** [comment]
> * Keep the code as `valid_index` rather than `invalid_index` so that we don't even up with double negatives.

Everywhere in the code the usage is `!valid_index`, which is why I preferred `invalid_index`. I renamed to `valid_index`  again.
> * Please leave the collections.deque code alone.  The author (me) aspires to keep this code not tightly coupled to the rest of 
the C API. The code is currently clean an doesn't need to be changed.

Done.
> * Are you _certain_ that inline functions always get inlined?  We know that macros do _always_ get inlined.  The `valid_index` macro currently gets used in performance critical code.

With inline functions I think we can never be certain the code gets inlined for all compilers and settings. I changed the static inline to a macro.

> * As Dennis says, this should not be part of the public API.  It is a performance trick that happens to currently be useful. As compilers get better, this would be done automatically and we would prefer the simple inline code rather than the extra layer of abstraction.

The `valid_index` is now private (part of ` Includes/internal/pycore_abstract.h`). If there is a more suitable location we can move it.

@rhettinger

**Comment by vstinner** [comment]
You should not use macro but a static inline functions, see:
* https://devguide.python.org/developer-workflow/c-api/index.html
* https://peps.python.org/pep-0670/

@rhettinger:
> Are you certain that inline functions always get inlined? We know that macros do always get inlined. The valid_index macro currently gets used in performance critical code.

See PEP 670 which answers to that and we approved by the Steering Council. A static inline function must be used.

File: Include/internal/pycore_abstract.h [hunk 1]
```diff
@@ -8,6 +8,16 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+/* Return 1 if the specified index is outside the range[0, limit) */
+#define valid_index(index, limit) ((size_t)index < (size_t)limit)
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]: Please convert it to a static inline function and rename it to _Py_is_valid_index().

**Review by vstinner — COMMENTED** [review]
I'm not very excited by a function only used for a micro-optimization, I'm fine with ``(i < 0 || i >= Py_SIZE(self))`` test. But I'm not against this change neither.

File: Include/internal/pycore_abstract.h (hunk 1 above)
Comment (eendebakpt) [reply Include/internal/pycore_abstract.h]: Done! I also updated the description of the PR in the first comment.

File: Include/internal/pycore_abstract.h [hunk 2]
```diff
@@ -8,6 +8,18 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+/// Return 1 if the specified index is outside the range [0, limit)
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]: I think that this comment is wrong. It returns non-zero if the index is *inside* the specific range, no?

Comment (eendebakpt) [reply Include/internal/pycore_abstract.h]: Good catch, I forgot to update after changing from `invalid_index` to `valid_index`

File: Python/bytecodes.c [hunk 3]
```diff
@@ -383,8 +383,7 @@ dummy_func(
             DEOPT_IF(!PyList_CheckExact(list), BINARY_SUBSCR);
 
             // Deopt unless 0 <= sub < PyList_Size(list)
-            Py_ssize_t signed_magnitude = Py_SIZE(sub);
-            DEOPT_IF(((size_t)signed_magnitude) > 1, BINARY_SUBSCR);
+            DEOPT_IF(_PyLong_Negative_or_multi_digit_int(sub), BINARY_SUBSCR);
```
Comment (vstinner) [inline Python/bytecodes.c]: Please revert changes related to _PyLong_Negative_or_multi_digit_int(), and write a separated PR adding _PyLong_Negative_or_multi_digit_int().

File: Include/internal/pycore_abstract.h [hunk 4]
```diff
@@ -8,6 +8,18 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+/// Return 1 if the specified index is in the range [0, limit)
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]: I'm always confused by range syntax, so I prefer a simple:

```suggestion
// Return 1 if 0 <= index < limit
```

File: Include/internal/pycore_abstract.h [hunk 5]
```diff
@@ -8,6 +8,18 @@ extern "C" {
...
+static inline int _Py_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
+{
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h:14]: You should mention that this function relies on twos-complement representation, and is not affected by -fwrapv, -fno-wrapv and -ftrapv compiler options of GCC and clang.

**Comment by vstinner** [comment]
I created https://github.com/python/cpython/issues/100008 to require two's complement to build Python.

File: Include/internal/pycore_long.h [hunk 6]
```diff
@@ -78,6 +78,14 @@ static inline PyObject* _PyLong_FromUnsignedChar(unsigned char i)
     return Py_NewRef((PyObject *)&_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS+i]);
 }
 
+/* Return 1 if the argument is negative or a multi-digit int */
+static inline int
+_PyLong_Negative_or_multi_digit_int(PyObject* sub) {
```
Comment (vstinner) [inline Include/internal/pycore_long.h]: Please revert this change since it's now discussed in a separated PR: PR #100064

Comment (eendebakpt) [reply Include/internal/pycore_long.h]: Done!

File: Include/internal/pycore_abstract.h [hunk 7]
```diff
@@ -8,6 +8,22 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+// Return 1 if 0 <= index < limit
+static inline int _Py_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]:
```suggestion
static inline int _Py_is_valid_index(Py_ssize_t index, Py_ssize_t limit)
```

File: Include/internal/pycore_abstract.h [hunk 8]
```diff
@@ -8,6 +8,22 @@ extern "C" {
...
+        to check whether i is in the range: 0 <= i < limit.
+        See:  Section 14.2 "Bounds Checking" in the Agner Fog
+        optimization manual found at:
+        https://www.agner.org/optimize/optimizing_cpp.pdf
+
+        The function relies on twos-complement representation, and is not
+        affected by -fwrapv, -fno-wrapv and -ftrapv compiler options
+        of GCC and clang
+    */
+    return (size_t)i < (size_t)limit;
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]:
```suggestion
    return (size_t)index < (size_t)limit;
```

File: Include/internal/pycore_abstract.h [hunk 9]
```diff
@@ -8,6 +8,22 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+// Return 1 if 0 <= index < limit
```
Comment (vstinner) [inline Include/internal/pycore_abstract.h]: The documentation uses "index", so I renamed the function parameter from "i" to "index". IMO it's a better name ;-)

File: Modules/arraymodule.c [hunk 10]
```diff
@@ -9,6 +9,7 @@
 
 #define PY_SSIZE_T_CLEAN
 #include "Python.h"
+#include "pycore_abstract.h"      // valid_index
```
Comment (vstinner) [inline Modules/arraymodule.c]:
```suggestion
#include "pycore_abstract.h"      // _Py_is_valid_index()
```

File: Modules/mmapmodule.c [hunk 11]
```diff
@@ -24,6 +24,7 @@
 
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
+#include "pycore_abstract.h"      // valid_index
```
Comment (vstinner) [inline Modules/mmapmodule.c]:
```suggestion
#include "pycore_abstract.h"      // _Py_is_valid_index()
```

File: Objects/codeobject.c [hunk 12]
```diff
@@ -3,6 +3,7 @@
 #include "Python.h"
 #include "opcode.h"
 #include "structmember.h"         // PyMemberDef
+#include "pycore_abstract.h"      // valid_index
```
Comment (vstinner) [inline Objects/codeobject.c]:
```suggestion
#include "pycore_abstract.h"      // _Py_is_valid_index()
```

File: Misc/NEWS.d/next/Core and Builtins/2022-11-25-09-34-57.gh-issue-99761.sRqH37.rst [hunk 13]
```diff
@@ -0,0 +1 @@
+Add optimized method ``_Py_is_valid_index`` to calculate valid indices
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2022-11-25-09-34-57.gh-issue-99761.sRqH37.rst]: Please remove the NEWS entry. This change doesn't affect Python users. I will add "skip news" label.

File: Objects/listobject.c [hunk 14]
```diff
@@ -2584,9 +2572,8 @@ list_index_impl(PyListObject *self, PyObject *value, Py_ssize_t start,
     }
     if (stop < 0) {
         stop += Py_SIZE(self);
-        if (stop < 0)
-            stop = 0;
```
Comment (vstinner) [inline Objects/listobject.c:2674]: I'm not 100% confident that this change is correct. If you consider that stop can no longer be negative, please remove the code an assertion: ``assert(stop >= 0);``.

**Comment by vstinner** [comment]
The overall change LGTM, but please address my two remaining change requests.

File: Objects/listobject.c (hunk 14 above)
Comment (eendebakpt) [reply Objects/listobject.c:2674]: Stop can still be negative at this point. But if `stop <0` then since `start >=0` the for loop is not evaluated. The same optimization is used in `tuple_index_impl`.

One more optimization in `tuple_index_impl` is to take the `i < Py_SIZE(self)` check out of the loop, but we cannot do that here since the length of the list can change (whereas the tuple is immutable)

**Comment by eendebakpt** [comment]
> The overall change LGTM, but please address my two remaining change requests.

I addressed the change requests. Based on the comments at #100064 I also removed the comment about the twos-complement.

File: Include/internal/pycore_abstract.h [hunk 15]
```diff
@@ -8,6 +8,21 @@ extern "C" {
...
+    /* The cast to size_t lets us use just a single comparison
+        to check whether i is in the range: 0 <= i < limit.
+        See:  Section 14.2 "Bounds Checking" in the Agner Fog
+        optimization manual found at:
+        https://www.agner.org/optimize/optimizing_cpp.pdf
+
+        The function is not affected by -fwrapv, -fno-wrapv and -ftrapv
+        compiler options of GCC and clang
+    */
+    return (size_t)index < (size_t)limit;
```
Comment (arhadthedev) [inline Include/internal/pycore_abstract.h:25]: Should we `assert(index >= 0); assert(limit >= 0);`?

Comment (eendebakpt) [reply Include/internal/pycore_abstract.h:25]: @arhadthedev We should indeed `assert(limit >= 0);`. The check is not valid for negative values of `limit`.
The `index` is allowed to be negative, in that case this method should return 0

**Comment by eendebakpt** [comment]
> The overall change LGTM, but please address my two remaining change requests.

@vstinner Could you have a look at the PR again?

**Review by arhadthedev — APPROVED** [review]
All earlier reviews seem to be addressed.

**Comment by serhiy-storchaka** [comment]
This idea was already discussed and rejected. See #72583.
